minkasi/needlet/needlet.py

```diff
@@ -562,98 +941,110 @@ def get_grid_dimless_2d(self, return_grid=False):
             return self.grid_dimless
 
 
-def new_map2wav(imaps, filters):
-    npix = imaps.shape[-2] * imaps.shape[-1]
-    weights = np.array([np.sum(f**2) / npix for f in filters])
-
-    fmap = np.fft.fft(
-        imaps, norm="backward"
-    )  # We're gonna do a hack-y thing where we force numpy to un normalize the ffts by calling norm=backwards going forwards and norm=forwards going backwards. We are imposing our own norm
-    fmap_npix = 1
-    for dim in np.shape(imaps):
-        fmap_npix *= dim
-    to_return = np.zeros(
-        (imaps.shape[0], filters.shape[0], imaps.shape[-2], imaps.shape[-1])
-    )
-    for i in range(len(imaps)):
-        for j in range(len(filters)):
-            fsmall = fmap[i]
-            fsmall *= filters[j] / (weights[i] ** 0.5 * fmap_npix)
-
-            to_return[i][j] = np.fft.ifft(fsmall, norm="forward").real
-    return to_return
-
-
-def map2wav_real(imaps, filters):
+def map2wav_real(
+    imaps: NDArray[np.floating],
+    filters: NDArray[np.floating],
+    n_filt: Optional[NDArray[np.floating]] = None,
+    print_time: Optional[bool] = False,
+) -> NDArray[np.floating]:
     """
     Transform from a regular map to a multimap of wavelet coefficients. Adapted from Joelles code + enmap.wavelets
 
     Parameters
     ----------
-    imap: np.array()
-        input map
-    basis:
-        needlet basis
-
+    imap : NDArray[np.floating]
+        Input map of the sky
+    filters : NDArray[np.floating]
+        Feedlet basis filters, also called the filter windows
+    n_filt : NDArray[np.floating] | None, default: None
+        Which filters to perform the coeficient evaluation for.
+        By default does all filters
+    print_time : bool, default: False
+        If true, prints the time taken to perform map2wav
     Returns
     -------
-    wmap: np.array
-        multimap of wavelet coefficients
+    wmap : NDArray[np.floating]
+        Multimap of wavelet coefficients with shape [nfilt, nx, ny]
     """
-
+    toc = time.time()
     if len(imaps.shape) == 2:
         imaps = np.expand_dims(imaps, axis=0)
+
     elif len(imaps.shape) != 3:
         print("Error: input map must have dim = 2 or 3")
         return
-    filtered_slices_real = []
 
-    npix = imaps.shape[-2] * imaps.shape[-1]
-    weights = np.array([np.sum(f**2) / npix for f in filters])
+    wmap = []
 
-    for i in range(len(imaps)):
+    if n_filt is None:
+        n_filt = range(len(imaps))
+
+    for i in n_filt:
         lightcone_ft = np.fft.fftn(np.fft.fftshift(imaps[i]))
```

Comment:
`np.fft` vs `mkfftw`? I don't remember where we landed in terms of speed here.

Reply:
So I did spend some time on this cause it's unsurprisingly by far the slowest thing when doing map2tod or tod2map. `pyfftw` is somewhat faster, maybe 20%. Currently `mkfftlib` only has 1D and 3D FFTs written, of course a 2D one could be written. Jon thinks the speedup form  `pyfftw` or `scipy` to FFTW is pretty marginal, he's probably right. I do have to play around with multithreading here, tho

Reply:
`mkfftw` is basically `pyfftw` but with some extra handling (append + flip) built in, so it not being much faster than `pyfftw` makes sense (and `pyfftw` is one of the scipy backends so same there). We could leave this like this for now and do a separate PR after some benchmarks to switch to the whole codebase to whatever fft backend is fastest (I know that `jax` uses the fft from `ducc` so maybe we try that library as well)

Reply:
Sorry I mean `pyfftw` is somewhat faster than `numpy`. I agree tho that benchmarking is for another PR.